addons/ui/fnc_progressBar.sqf

```diff
@@ -0,0 +1,172 @@
+
+if (!isNull _display) then {
+    // run failure code on previous progress bar
+    private _arguments = _display getVariable QGVAR(arguments);
+    private _onFailure = _display getVariable QGVAR(onFailure);
+
+    private _startTime = _display getVariable QGVAR(startTime);
+    private _totalTime = _display getVariable QGVAR(totalTime);
+    private _elapsedTime = (CBA_missionTime - _startTime) min _totalTime;
+
+    _display closeDisplay 0;
+    [_onFailure, [_arguments, false, _elapsedTime, _totalTime]] call CBA_fnc_execNextFrame;
+};
+
+// create new progress bar
+private _blockMouse = _blockInput || _allowClose;
+
+if (_blockMouse) then {
+    _display = (uiNamespace getVariable "RscDisplayMission") createDisplay QGVAR(ProgressBar);
+
+    _display displayAddEventHandler ["KeyDown", {
+        params ["_display", "_key", "_shift", "_control", "_alt"];
+
+        private _arguments = _display getVariable QGVAR(arguments);
+        private _onFailure = _display getVariable QGVAR(onFailure);
+        private _allowClose = _display getVariable QGVAR(allowClose);
+        private _blockInput = _display getVariable QGVAR(blockInput);
+
+        if (_allowClose && {_key isEqualTo DIK_ESCAPE}) then {
+            private _startTime = _display getVariable QGVAR(startTime);
+            private _totalTime = _display getVariable QGVAR(totalTime);
+            private _elapsedTime = (CBA_missionTime - _startTime) min _totalTime;
+
+            _display closeDisplay 0;
+            [_onFailure, [_arguments, false, _elapsedTime, _totalTime]] call CBA_fnc_execNextFrame;
+
+            _blockInput = true;
+        };
+
+        _blockInput
+    }];
+} else {
+    QGVAR(ProgressBar) cutRsc [QGVAR(ProgressBar), "PLAIN"];
+    _display = uiNamespace getVariable QGVAR(ProgressBar);
+};
+
+private _fnc_check = {
+    private _display = uiNamespace getVariable [QGVAR(ProgressBar), displayNull];
+
+    if (isNull _display) exitWith {
+        removeMissionEventHandler ["Draw3D", _thisEventHandler];
+    };
+
+    private _arguments = _display getVariable QGVAR(arguments);
+    private _condition = _display getVariable QGVAR(condition);
+
+    private _startTime = _display getVariable QGVAR(startTime);
+    private _totalTime = _display getVariable QGVAR(totalTime);
+    private _elapsedTime = (CBA_missionTime - _startTime) min _totalTime;
+
+    private _continue = [[_arguments, true, _elapsedTime, _totalTime], _condition] call {
+        private ["_display", "_arguments", "_condition", "_startTime", "_totalTime", "_elapsedTime"];
+        _this#0 call _this#1;
+    };
+
+    private _onSuccess = _display getVariable QGVAR(onSuccess);
+    private _onFailure = _display getVariable QGVAR(onFailure);
+
+    if (!_continue) exitWith {
+        _display closeDisplay 0;
+        [_onFailure, [_arguments, false, _elapsedTime, _totalTime]] call CBA_fnc_execNextFrame;
+    };
+
+    if (_elapsedTime > _totalTime) exitWith {
```

Comment:
`_elapsedTime = (CBA_missionTime - _startTime) min _totalTime;`
so elapsed can never be greater than total